junifer/pipeline/workdir_manager.py

```diff
@@ -0,0 +1,112 @@
+"""Provide a work directory manager class to be used by pipeline components."""
+
+# Authors: Synchon Mandal <s.mandal@fz-juelich.de>
+#          Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
+
+import shutil
+import tempfile
+from pathlib import Path
+from typing import Optional, Union
+
+from ..utils import logger
+from .singleton import singleton
+
+
+@singleton
+class WorkDirManager:
+    """Class for working directory manager.
+
+    This class is a singleton and is used for managing temporary and working
+    directories used across the pipeline by datagrabbers, preprocessors,
+    markers and so on. It maintains a single super-directory and provides
+    directories on-demand and cleans after itself thus keeping the user
+    filesystem clean.
+
+
+    Parameters
+    ----------
+    workdir : str or pathlib.Path, optional
+        The path to the super-directory. If None, "TMPDIR/junifer" is used
+        where TMPDIR is the platform-dependent temporary directory.
+
+    """
+
+    def __init__(self, workdir: Optional[Union[str, Path]] = None) -> None:
```

Comment:
Since it's a singleton, any intialization should take care of cleaning up a previous one.

e.g:

```
WorkdirManager("/tmp/workdir1")
# Do some stuff that creates tempdirs
WorkdirManager("/tmp/workdir2")  # This should cleanup.
```

However, if a user does use it wrongly, it might cleanup in the middle.

So I would: 
1) add a `cleanup=False` parameter to the init
2) Raise an error if needs cleanup but cleanup is False

Reply:
I've made the changes as you proposed. The problem is that the `__init__()` doesn't fire the second time you try to create an instance as it's a singleton, so the previous instance is returned.

Reply:
Ok, so how can we fix it?

Reply:
The only way I know of interacting with singletons in principle is to use getters and setters.

Reply:
lets do that then